source/jetstream/iterator.go

```diff
@@ -0,0 +1,276 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jetstream
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/nats-io/nats.go"
+)
+
+// heartbeatTimeout is a default heartbeat timeout for push consumers.
+const heartbeatTimeout = 2 * time.Second
+
+// Iterator is a iterator for JetStream communication model.
+// It receives message from NATS JetStream.
+type Iterator struct {
+	sync.Mutex
+
+	conn          *nats.Conn
+	messages      chan *nats.Msg
+	unackMessages []*nats.Msg
+	jetstream     nats.JetStreamContext
+	consumerInfo  *nats.ConsumerInfo
+	subscription  *nats.Subscription
+}
+
+// IteratorParams contains incoming params for the NewIterator function.
+type IteratorParams struct {
+	Conn          *nats.Conn
+	BufferSize    int
+	Durable       string
+	Stream        string
+	Subject       string
+	SDKPosition   sdk.Position
+	DeliverPolicy nats.DeliverPolicy
+	AckPolicy     nats.AckPolicy
+}
+
+// NewIterator creates new instance of the Iterator.
+func NewIterator(ctx context.Context, params IteratorParams) (*Iterator, error) {
+	jetstream, err := params.Conn.JetStream()
+	if err != nil {
+		return nil, fmt.Errorf("get jetstream context: %w", err)
+	}
+
+	consumerConfig, err := getConsumerConfig(params)
+	if err != nil {
+		return nil, fmt.Errorf("get consumer config: %w", err)
+	}
+
+	consumerInfo, err := jetstream.AddConsumer(params.Stream, consumerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("add jetstream consumer: %w", err)
+	}
+
+	messages := make(chan *nats.Msg, params.BufferSize)
+
+	subscription, err := jetstream.ChanSubscribe(params.Subject, messages,
+		nats.Durable(consumerInfo.Config.Durable),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("chan subscribe: %w", err)
+	}
+
+	return &Iterator{
+		conn:          params.Conn,
+		messages:      messages,
+		unackMessages: make([]*nats.Msg, 0),
+		jetstream:     jetstream,
+		consumerInfo:  consumerInfo,
+		subscription:  subscription,
+	}, nil
+}
+
+// HasNext checks is the iterator has messages.
+func (i *Iterator) HasNext(ctx context.Context) bool {
+	return len(i.messages) > 0
+}
+
+// Next returns the next record from the underlying messages channel.
+// It also appends messages to a unackMessages slice if the AckPolicy is not equal to AckNonePolicy.
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
```

Comment:
Do we need add for loop?

Reply:
I think we don't need it here. Cause we just need to read one record from the channel